src/services/Redis.ts

```diff
@@ -1,63 +1,110 @@
-            expire: promisify(this.client.expire).bind(this.client),
-            expireAt: promisify(this.client.expireat).bind(this.client),
-            get: promisify(this.client.get).bind(this.client),
-            set: promisify(this.client.set).bind(this.client),
-        };
+    public constructor(config?: ClientOpts, ephemeral?: boolean) {
+        this.ephemeral = ephemeral;
+        this.config = config;
+        this.listenersRegistered = false;
     }
 
     /**
      * Retrieves the keys value from the cache.
      */
     public async get(key: string): Promise<string> {
-        return this.commands.get(key);
+        let data = null;
+        this.execute(methodLabels.get, key, (error, result) => {
+            data = result;
+        });
+        return data;
     }
 
     /**
      * Persists the key/value pair in the cache,
      * optionally setting it to expire at a particular time or in a given number of seconds.
      */
     public async put(key: string, value: string, expire?: Expire): Promise<void> {
-        await this.commands.set(key, value);
+        this.execute(methodLabels.set, key, value, () => {
+            if (expire) {
+                if (expire.at) {
+                    this.client.expireat(key, expire.at);
+                }
 
-        if (expire) {
-            if (expire.at) {
-                await this.commands.expireAt(key, expire.at);
+                if (expire.in) {
+                    this.client.expire(key, expire.in);
+                }
             }
+        });
+    }
+
+    /**
+     * Removes the key/value pair from the cache.
+     */
+    public async remove(key: string): Promise<void> {
+        this.execute(methodLabels.del, key);
+    }
 
-            if (expire.in) {
-                await this.commands.expire(key, expire.in);
+    /**
+     * Creates new Redis instance if one does not already exist or if is configured to be ephemeral.
+     * If configured to be ephemeral redis actions will have event listeners on them.
+     */
+    private getClient(): RedisClient {
+        if (this.ephemeral || !this.client) {
+            this.client = createClient(this.config);
+            if (this.ephemeral && !this.listenersRegistered) {
+                this.registerListeners();
             }
         }
+        return this.client;
     }
 
     /**
-     * Removes the key/value pair from the cache.
+     * Attaches listeners to Redis commands.
      */
-    public async remove(key: string): Promise<void> {
-        await this.commands.del(key);
+    private registerListeners() {
+        Object.keys(methodLabels).forEach(label => {
+            this.callbacks.addListener(label, () => {
+                this.client.quit();
+            });
+        });
+
+        this.listenersRegistered = true;
+    }
+
+    /**
+     * Executes Redis command and emits event if ephemeral.
+     */
+    private execute(method: string, ...args) {
```

Comment:
we should use enum here instead of string

Reply:
👍 done